### PR TITLE
feat: add SQL parsing time-consuming metric

### PR DESCRIPTION
**Pull request by yanni97**
- Add tracking points for SQL parsing time consumption, and use these tracking points as performance analysis data support

### CODE REVIEW

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java [hunk 1]
```diff
@@ -151,13 +164,32 @@ public static TableSchema parse(String sql) {
...
         return statement.getClass().getSimpleName();
     }
+
+    private void recordParseTimeAndLength(long duration, int sqlLength) {
+        if (CollectionUtils.isEmpty(metricListenerList)) {
+            return;
+        }
+        long durationMillis = nanosToMillis(duration);
+        for (MetricListener metricListener : metricListenerList) {
+            metricListener.recordTime(SQL_PARSE_TIME_METRIC_NAME, Collections.emptyMap(), durationMillis);
```
Comment (pangdayuan1) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:185]: Hiding points requires tags, Can refer to
com.arextest.storage.metric.AgentWorkingMetricService#recordEntryTime

Comment (yanni97) [reply arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:185]: Using appId as tag?

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java [hunk 2]
```diff
@@ -133,31 +147,48 @@ public static List<String> parseTableNames(String operationName) {
...
+            Set<String> tableNameList = new TablesNamesFinder().getTables(statement);
+            // sort table nam, Sort in lexicographical order
             if (CollectionUtils.isNotEmpty(tableNameList)) {
-                Collections.sort(tableNameList);
+                tableNameList = new TreeSet<>(tableNameList);
             }
             tableSchema.setTableNames(tableNameList);
         } catch (Throwable e) {
-            LOGGER.warn("parse sql error, sql: {}", sql, e);
+            return tableSchema;
```
Comment (pangdayuan1) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:167]: I have a question. When catching Throwable exceptions and oom occurs, will the log still be printed?

Comment (yanni97) [reply arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:167]: Won't

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java [hunk 3]
```diff
@@ -133,31 +147,48 @@ public static List<String> parseTableNames(String operationName) {
...
         TableSchema tableSchema = new TableSchema();
         try {
             sql = PATTERN.matcher(sql).replaceAll(" ");
 
             Statement statement = CCJSqlParserUtil.parse(sql);
             tableSchema.setAction(getAction(statement));
 
-            List<String> tableNameList = new TablesNamesFinder().getTableList(statement);
-            // sort table name
+            Set<String> tableNameList = new TablesNamesFinder().getTables(statement);
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:160]: `tableNameList` -> `tableNameSet`

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java [hunk 4]
```diff
@@ -133,31 +147,48 @@ public static List<String> parseTableNames(String operationName) {
...
+            // sort table nam, Sort in lexicographical order
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:161]: // sort table name

**Review by lucas-myx — DISMISSED** [review]

**Review by pangdayuan1 — DISMISSED** [review]

**Review by lucas-myx — APPROVED** [review]

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java [hunk 5]
```diff
@@ -149,15 +166,32 @@ public static TableSchema parse(String sql) {
...
     }
+
+    private void recordParseTime(long duration, String sql, String appId) {
+        if (CollectionUtils.isEmpty(metricListenerList)) {
+            return;
+        }
+        Map<String, String> tags = Collections.singletonMap(CLIENT_APP_ID, appId);
+        for (MetricListener metricListener : metricListenerList) {
+            metricListener.recordTime(SQL_PARSE_TIME_METRIC_NAME, tags, duration);
+            int parseTimeThreshold = applicationDefaultConfig.getConfigAsInt(SQL_PARSE_DURATION_THRESHOLD, SQL_PARSE_DURATION_THRESHOLD_DEFAULT);
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:191]: put configuration fetch outside the loop

**Review by pangdayuan1 — DISMISSED** [review]

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java [hunk 6]
```diff
@@ -149,15 +166,32 @@ public static TableSchema parse(String sql) {
...
+    private void recordParseTime(long duration, String sql, String appId) {
+        if (CollectionUtils.isEmpty(metricListenerList)) {
+            return;
+        }
+        int parseTimeThreshold = applicationDefaultConfig.getConfigAsInt(SQL_PARSE_DURATION_THRESHOLD, SQL_PARSE_DURATION_THRESHOLD_DEFAULT);
+        Map<String, String> tags = Collections.singletonMap(CLIENT_APP_ID, appId);
+        for (MetricListener metricListener : metricListenerList) {
+            metricListener.recordTime(SQL_PARSE_TIME_METRIC_NAME, tags, duration);
+            if (duration > parseTimeThreshold) {
+                LOGGER.warn("[[title=sqlParse]]the actual parsing time:{} exceeds the set threshold:{}, sql: {}", duration, parseTimeThreshold, sql);
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/DatabaseParseService.java:193]: put logger outside the loop, too

File: arex-storage-web-api/src/main/java/com/arextest/storage/model/TableSchema.java [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 
 import java.util.List;
+import java.util.Set;
+
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/model/TableSchema.java:6]: Remove unsed import, leave this class unchanged.

**Review by mr3 — DISMISSED** [review]

**Review by pangdayuan1 — APPROVED** [review]

**Review by mr3 — APPROVED** [review]

**Review by coryhh — APPROVED** [review]

**Pull request merged**